source/common/filter/config_discovery_impl.h

```diff
@@ -125,17 +124,41 @@ class DynamicFilterConfigProviderImpl : public DynamicFilterConfigProviderImplBa
   }
 
 private:
+  FactoryCb instantiateFilterFactory(const Protobuf::Message& message) const {
+    auto* factory = Registry::FactoryRegistry<Factory>::getFactoryByType(message.GetTypeName());
+    return factory->createFilterFactoryFromProto(message, stat_prefix_, factory_context_);
+  }
+
+  void update(absl::optional<FactoryCb> config, Config::ConfigAppliedCb applied_on_all_threads) {
+    tls_.runOnAllThreads([config](OptRef<ThreadLocalConfig> tls) { tls->config_ = config; },
+                         [main_config = main_config_, config, applied_on_all_threads]() {
+                           // This happens after all workers have discarded the previous config so
+                           // it can be safely deleted on the main thread by an update with the new
+                           // config.
+                           main_config->current_config_ = config;
+                           if (applied_on_all_threads) {
+                             applied_on_all_threads();
+                           }
+                         });
+  }
+
   struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
     ThreadLocalConfig() : config_{absl::nullopt} {}
     absl::optional<FactoryCb> config_{};
   };
 
   // Currently applied configuration to ensure that the main thread deletes the last reference to
-  // it.
-  absl::optional<FactoryCb> current_config_{absl::nullopt};
+  // it. Filter factories may hold their own thread local storage which is required to be deleted
+  // on the main thread.
+  struct MainConfig {
+    absl::optional<FactoryCb> current_config_{absl::nullopt};
+  };
+
+  const std::string stat_prefix_;
+  Server::Configuration::FactoryContext& factory_context_;
+  std::shared_ptr<MainConfig> main_config_;
```

Comment:
I'm not certain that shared_ptr makes a difference in this case.
IIUC the only update is on the pointed object (`current_config_`), which would be overrun whether this is a shared pointer or not.

Reply:
During destruction, `main_config` is shared with the `runOnAllThreads` callbacks which guarantees that the last reference is deleted on the main thread. That ensures that the inner config is always deleted on the main thread, which is necessary if the inner config has its own TLS.

Reply:
I see, thanks for clarification.